### PR TITLE
macho: extend CodeSignature to accept entitlements

### PR DESCRIPTION
**Pull request by kubkon**
With this change, we can now bake in entitlements into the binary.
Additionally, I see this as the first step towards full code signature
support which includes baking in Apple issued certificates for
redistribution, etc.

### CODE REVIEW

File: src/main.zig [hunk 1]
```diff
@@ -1036,6 +1037,10 @@ fn buildOutputType(
                         } else {
                             enable_link_snapshots = true;
                         }
+                    } else if (mem.eql(u8, arg, "--entitlements")) {
```
Comment (andrewrk) [inline src/main.zig:1040]: don't forget to update the corresponding usage text above this function

**Review by andrewrk — CHANGES_REQUESTED** [review]
One minor thing, and a question: what is the clang / apple tooling way to specify an entitlements file? I didn't find an `--entitlements` command line parameter for clang. Does it have a corresponding flag? If so we should adjust `zig cc` to recognize it too.

**Comment by kubkon** [comment]
> One minor thing, and a question: what is the clang / apple tooling way to specify an entitlements file? I didn't find an `--entitlements` command line parameter for clang. Does it have a corresponding flag? If so we should adjust `zig cc` to recognize it too.

It doesn't, and most likely never will. Linker only creates the minimal adhoc signature, while `codesign` tool is used as part of the full Xcode toolchain to embed advanced code signature objects which include entitlements, etc.

File: src/Compilation.zig [hunk 2]
```diff
@@ -2302,6 +2305,7 @@ fn addNonIncrementalStuffToCacheManifest(comp: *Compilation, man: *Cache.Manifes
 
     man.hash.addListOfBytes(comp.clang_argv);
 
+    man.hash.addOptionalBytes(comp.bin_file.options.entitlements);
```
Comment (andrewrk) [inline src/Compilation.zig:2308]: there's a "Mach-O specific stuff" section about 30 lines below. Also this is a file so it needs to be added to the manifest as a file dependency:

```suggestion
    man.hash.addOptionalFile(comp.bin_file.options.entitlements);
```

Same deal with the other cache API usage in the linker.

File: src/main.zig [hunk 3]
```diff
@@ -433,6 +433,7 @@ const usage_build_generic =
     \\  -framework [name]              (Darwin) link against framework
     \\  -F[dir]                        (Darwin) add search path for frameworks
     \\  -install_name=[value]          (Darwin) add dylib's install name
+    \\  --entitlements=[path]          (Darwin) add path to entitlements file for embedding in code signature
```
Comment (andrewrk) [inline src/main.zig:436]: Looks like it does not parse with an equal sign:

```suggestion
    \\  --entitlements [path]          (Darwin) add path to entitlements file for embedding in code signature
```

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Review by andrewrk — APPROVED** [review]

**Pull request merged**